packages/backend-app-api/src/wiring/BackendInitializer.ts

```diff
@@ -170,18 +172,26 @@ export class BackendInitializer {
     }
 
     // Initialize all root scoped services
+    const featureRegistrationObservers: BackendFeatureRegistrationObserver[] =
+      [];
     for (const ref of this.#serviceRegistry.getServiceRefs()) {
       if (ref.scope === 'root') {
-        await this.#serviceRegistry.get(ref, 'root');
+        const service = await this.#serviceRegistry.get(ref, 'root');
+        if (service instanceof BackendFeatureRegistrationObserver) {
```

Comment:
We pretty strongly favour composition over inheritance. I'd prefer to not _force_ everyone to use classes and inheritance when all they want to do is to listen to an event. And the notion of "just implement this interface and you'll implicitly get some features" unfortunately isn't very well aligned with how we do things elsewhere.

(as a personal note, every usage of the `abstract` and `instanceof` keywords stands out, at least in the way I review things, and usually leads to comments)

Would it make sense to instead make an actual

```ts
/**
 * @public
 */
export interface FeatureRegistryService {
  onRegistrationsChanged(callback: () => void): { unregister: () => void };
  getRegistrations(): BackendFeatureRegistration[];
}
```

with a serviceRef and all? Could also choose to send the new array of registrations directly to the callback, or use an observable member like `registration$(): Observable<BackendFeatureRegistration[]>` like we do in [some other services](https://github.com/backstage/backstage/blob/2bd94c454e82c955450cb791a8f54c0da7adc7af/packages/core-app-api/src/apis/implementations/auth/saml/SamlAuth.ts#L81), but that's all a matter of taste how you prefer to build it.

Then your test above would be making a mock of this and then asserting that some registrations change callback was called twice with the expected arguments or so.

@Rugvip ping here, if you've been in discussions around this interface before or so

Reply:
I get your comments. I came back to such a minimal / internal design after @Rugvip comments of previous PR https://github.com/backstage/backstage/pull/22332.

Reply:
It seems to me that what you proposed is somehow what I did in the [previous PR](https://github.com/backstage/backstage/pull/22332).
This PR is mainly a cut in the previous PR.